### PR TITLE
[SPARK-11916][SQL]Expression TRIM/LTRIM/RTRIM to support specific trim word

### PR DESCRIPTION
**Pull request by adrian-wang**
supports expressions like `trim('xxxabcxxx', 'x')`

### CODE REVIEW

**Comment by adrian-wang** [comment]
The failure results from old TRIM would trim characters whose ascii is in [0, 20], but actually we should only trim ' '. This is to stay consistent with Hive 1.2.1.

**Comment by adrian-wang** [comment]
cc @chenghao-intel

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 1]
```diff
@@ -343,47 +343,71 @@ case class FindInSet(left: Expression, right: Expression) extends BinaryExpressi
 }
 
 /**
- * A function that trim the spaces from both ends for the specified string.
+ * A function that trims the given character (by default space) from both ends for given string.
  */
-case class StringTrim(child: Expression)
-  extends UnaryExpression with String2StringExpression {
+case class StringTrim(left: Expression, right: Expression)
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: Nit: I am wondering if we can use more meaningful name, like `expr` and `trimChar` for constructor arguments.
And also add `def left = expr` and `def right = trimChar`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala [hunk 2]
```diff
@@ -32,6 +32,7 @@ object Literal {
     case d: Double => Literal(d, DoubleType)
     case f: Float => Literal(f, FloatType)
     case b: Byte => Literal(b, ByteType)
+    case c: Char => Literal(c.toByte, ByteType)
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/literals.scala]: This is probably not right in the casting, `Char` is actually 2 bytes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala [hunk 3]
```diff
@@ -343,47 +343,71 @@ case class FindInSet(left: Expression, right: Expression) extends BinaryExpressi
...
-  extends UnaryExpression with String2StringExpression {
+case class StringTrimRight(left: Expression, right: Expression)
+  extends BinaryExpression with ImplicitCastInputTypes {
+
+  def this(token: Expression) = {
+    this(token, Literal(' '))
+  }
+
+  override def dataType: DataType = StringType
+  override def inputTypes: Seq[DataType] = Seq(StringType, ByteType)
```
Comment (chenghao-intel) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala]: Using the `StringType` also for the second argument? e.g. In MySQL, it supports case like:
`SELECT TRIM(LEADING 'leading' FROM 'leadingtext' )`  => `text`

**Pull request closed without merging**